src/angular/components/change-password.component.ts

```diff
@@ -0,0 +1,267 @@
+                }
+
+                await this.formPromise;
+
+                this.platformUtilsService.showToast('success', this.i18nService.t('masterPasswordChanged'),
+                    this.i18nService.t('logBackIn'));
+                this.messagingService.send('logout');
+            } catch {
+                this.platformUtilsService.showToast('error', null, this.i18nService.t('errorOccurred'));
+            }
+        }
+    }
+
+    updatePasswordStrength() {
+        if (this.masterPasswordStrengthTimeout != null) {
+            clearTimeout(this.masterPasswordStrengthTimeout);
+        }
+        this.masterPasswordStrengthTimeout = setTimeout(() => {
+            const strengthResult = this.passwordGenerationService.passwordStrength(this.newMasterPassword,
+                this.getPasswordStrengthUserInput());
+            this.masterPasswordScore = strengthResult == null ? null : strengthResult.score;
+        }, 300);
+    }
+
+    async rotateEncKeyClicked() {
+        if (this.rotateEncKey) {
+            const ciphers = await this.cipherService.getAllDecrypted();
+            let hasOldAttachments = false;
+            if (ciphers != null) {
+                for (let i = 0; i < ciphers.length; i++) {
+                    if (ciphers[i].organizationId == null && ciphers[i].hasOldAttachments) {
+                        hasOldAttachments = true;
+                        break;
+                    }
+                }
+            }
+
+            if (hasOldAttachments) {
+                const learnMore = await this.platformUtilsService.showDialog(
+                    this.i18nService.t('oldAttachmentsNeedFixDesc'), null,
+                    this.i18nService.t('learnMore'), this.i18nService.t('close'), 'warning');
+                if (learnMore) {
+                    this.platformUtilsService.launchUri(
+                        'https://help.bitwarden.com/article/attachments/#fixing-old-attachments');
+                }
+                this.rotateEncKey = false;
+                return;
+            }
+
+            const result = await this.platformUtilsService.showDialog(
+                this.i18nService.t('updateEncryptionKeyWarning') + ' ' +
+                this.i18nService.t('rotateEncKeyConfirmation'), this.i18nService.t('rotateEncKeyTitle'),
+                this.i18nService.t('yes'), this.i18nService.t('no'), 'warning');
+            if (!result) {
+                this.rotateEncKey = false;
+            }
+        }
+    }
+
+    private getPasswordStrengthUserInput() {
+        let userInput: string[] = [];
+        const atPosition = this.email.indexOf('@');
+        if (atPosition > -1) {
+            userInput = userInput.concat(this.email.substr(0, atPosition).trim().toLowerCase().split(/[^A-Za-z0-9]/));
+        }
+        return userInput;
+    }
+
+    private async updateKey(key: SymmetricCryptoKey, masterPasswordHash: string) {
```

Comment:
Same. seems specific to a particular component

Reply:
Refactored and updated.